### PR TITLE
Draft new runtime package

### PR DESCRIPTION
**Pull request by mdrohmann**
https://www.pivotaltracker.com/story/show/176853423

### CODE REVIEW

**Comment by mdrohmann** [comment]
I tried to be as close as possible to our draft, but introduced a few changes, as I started to scaffolding the code.

Most notably I introduced an interface for all backend functions such that we can substitute it with a mock for testing. I think that could be very powerful, as the `Setup.InstallRuntime()` function is quite complex and would profit from a good unit test IMO.

The general structure is all document in `pkg/platform/runtime2/doc.go`: `go doc pkg/platform/runtime2`, and I tried to fill out the `Setup.InstallRuntime()` a bit to demonstrate how the general control flow could look like.

File: pkg/platform/runtime2/api/api.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package api
```
[the remark this answers is not in the document]
Comment (mdrohmann) [reply pkg/platform/runtime2/api/api.go:1]: Well, it actually is more of  a `model` in our current terminology.  Should it live here though, or in `pkg/platform/model`?

File: pkg/platform/runtime2/doc.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+- setup/setup_test.go: tests the entire set up of a runtime based on a mocked
+API client. This is the most complicated part, and it involves some
+asynchronous operations. So, it will be nice to have some unit-tests
+available, to test some edge cases, especially w.r.t. message handling.
+- artifact/artifact_test.go: Tests to ensure that we can parse the returned
+Recipe structure correctly.
+- api/client/default_test.go: Here we could add some very focused integration
+tests, that should fail if the backend changes in an incompatible way.
+*/
+package runtime
```
[the remark this answers is not in the document]
Comment (mdrohmann) [reply pkg/platform/runtime2/doc.go:98]: Writing the `doc.go` file, I was also surprised how little formatting options there were.  I will transform it into a ARCHITECTURE.md.

File: pkg/platform/runtime2/runtime.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+func IsNotInstalledError(err error) bool {
+	return errors.Is(err, ErrNotInstalled)
+}
+
+// NewRuntime creates a new runtime.
+//
+// The actual type is determined by the buildEngine file.
+// Usually, this function is not called directly but via `setup.Setup.InstalledRuntime()`.
+// If the runtime cannot be initialized a NotInstalledError is returned.
+func NewRuntime(proj *project.Project) (common.Runtimer, error) {
```
[the remark this answers is not in the document]
Comment (mdrohmann) [reply pkg/platform/runtime2/runtime.go:25]: That's a valid point.  It is probably save to remove this function completely...

**Comment by mdrohmann** [comment]
I addressed some @Naatan and @daved 's suggestions.  The documentation is now [in a markdown file](https://github.com/ActiveState/cli/blob/fa2b4e35669a72bd09c91a2604babd12089e3f00/pkg/platform/runtime2/ARCHITECTURE.md).

**Review by daved — CHANGES_REQUESTED** [review]
The usefulness of the strategy pattern (what I've been calling "drivers" in conversation) to prevent some awkward use of interfaces is still being missed out on. I worked on some organization that should more clearly demonstrate the benefits: https://github.com/ActiveState/cli/compare/dgreen/martind-runtime_refactor-176853423?expand=1

For an additional point of reference, a "Runtimer" interface would be defined by calling packages (like controllers/runners), not the runtime package itself. (Yes, the EnvProvider interface is probably identical to Runtimer, but the name and purpose is defined by/within the runtime package scope, and it's incidental that runtime.Runtime just forwards the call.)

Edit to add: There are a number of areas that can be straightened out more, but, for now, I don't have enough working knowledge of the needs to spend any more time on this.

File: pkg/platform/runtime2/setup/setup.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+// NewSetup returns a new Setup instance that can install a Runtime locally on the machine.
+func NewSetup(project *project.Project, msgHandler impl.MessageHandler) *Setup {
+	return NewSetupWithAPI(project, msgHandler, client.NewDefault())
+}
+
+// NewSetupWithAPI returns a new Setup instance with a customized API client eg., for testing purposes
+func NewSetupWithAPI(project *project.Project, msgHandler impl.MessageHandler, api model.ClientProvider) *Setup {
+	panic("implement me")
+}
```
[the remark this answers is not in the document]
Comment (daved) [reply pkg/platform/runtime2/setup/setup.go:32]: I included this in my example. Marking this and the other conversation as resolved for now.

File: pkg/platform/runtime2/api/api.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Naatan) [reply pkg/platform/runtime2/api/api.go]: `model` is for structuring requests and processing responses, it shouldn't deal with api configuration or actually processing the requests/responses on the network level.

I'd say most of what I'm seeing here should live in the existing `api` directory. In fact, aren't we already covering this there?

[the remark this answers is not in the document]
Comment (mdrohmann) [reply pkg/platform/runtime2/api/api.go]: Yes, `api` was an unfortunate name for this probably.  I never intended the `ClientProvider` to actually deal with requests on the network level.  In its default installation, it would just be a thin wrapper around all the APIs that we need to use here: `graphql`, `headchef` and `buildlogstream`.  But as it is three of them, I find this concept warranted.  Its main purpose is to facilitate the `client.Mock` implementation that fulfills the `ClientProvider` interface without doing any network requests (not even to a mocked http server).

In my opinion this is more of a `model` therefore, as in the implementations we (can) do some post-processing of API calls.

**Comment by mdrohmann** [comment]
> The usefulness of the strategy pattern (what I've been calling "drivers" in conversation) to prevent some awkward use of interfaces is still being missed out on. I worked on some organization that should more clearly demonstrate the benefits: https://github.com/ActiveState/cli/compare/dgreen/martind-runtime_refactor-176853423?expand=1
> 
> For an additional point of reference, a "Runtimer" interface would be defined by calling packages (like controllers/runners), not the runtime package itself. (Yes, the EnvProvider interface is probably identical to Runtimer, but the name and purpose is defined by/within the runtime package scope, and it's incidental that runtime.Runtime just forwards the call.)
> 
> Edit to add: There are a number of areas that can be straightened out more, but, for now, I don't have enough working knowledge of the needs to spend any more time on this.

I like the changes, except for maybe the move of the `apiclient` to a `runtime2` sibling directory, because I believe it is (a) more of a `model` than an `api` (c.f., [here](https://github.com/ActiveState/cli/pull/1252#discussion_r575403397), and (b) I believe this is so specific to the `runtime2` use-case that is better to organize it relative to that directory.  But that is maybe more of a personal taste.
